base/src/main/java/org/aya/core/pat/Pat.java

```diff
@@ -225,6 +227,50 @@ record End(boolean isRight, boolean explicit) implements Pat {
     }
   }
 
+  /** TODO[literal]: literal type needs meta-solving for first-class patterns */
+  record ShapedInt(
+    @Override int repr,
+    @Override @NotNull AyaShape shape,
+    @NotNull CallTerm.Data type,
+    boolean explicit
+  ) implements Pat, Shaped.Inductively<Pat> {
+    @Override public @NotNull Expr toExpr(@NotNull SourcePos pos) {
+      return new Expr.LitIntExpr(pos, repr);
+    }
+
+    @Override public @NotNull Pat rename(@NotNull Subst subst, @NotNull LocalCtx localCtx, boolean explicit) {
+      return this;
+    }
+
+    @Override public @NotNull Pat zonk(@NotNull Tycker tycker) {
+      return this;
+    }
+
+    @Override public @NotNull Pat inline() {
+      return this;
+    }
+
+    @Override public void storeBindings(@NotNull LocalCtx localCtx) {
+      // do nothing
+    }
+
+    @Override public @NotNull Pat makeZero(@NotNull CtorDef zero) {
+      return new Pat.Ctor(explicit, zero.ref, ImmutableSeq.empty(), type);
```

Comment:
Maybe make `type` a `Term` and report if `type` is not data call

Reply:
Yes I did this before but I cannot come up with a case that the pattern's type is a hole. Maybe we should keep the code consistent between literal Terms and Pats.

Reply:
Well Idk either, will `def f (a b : _) : a + b = b + a` work? But it seems that the types will be solved before checking the patterns. Hmmm...

Reply:
> Well Idk either, will `def f (a b : _) : a + b = b + a` work?

Currently this code does not tyck because the hole was unsolved (my test case follows)
```
def f (a b : _) : Nat
  | 0, 0 => 0
  | a, b => a
```

> But it seems that the types will be solved before checking the patterns

Yes. But if the types are unsolved, no patterns will be checked further.

Reply:
If we make the `type` of Pat a `Term`, we may add some yet-unreachable code. So I propose to do this when we need to solve the type of patterns.